File: apps/mcp/src/audit.ts
import { existsSync, lstatSync, mkdirSync, realpathSync, statSync } from 'node:fs';
import { join, resolve } from 'node:path';
import {
  createBoundEventAppender,
  sanitizeEvent,
  validateEvent,
  type CollaborationEvent,
  type CollaborationEventType,
  type ObjectKind,
} from '@openslack/collaboration';
import {
  canonicalizeGovernedJson,
  type GovernedJsonValue,
  type GovernedPlanAuditEvent,
  type GovernedPlanAuditSink,
} from '@openslack/operator';

const EVENT_TYPES = Object.freeze({
  'plan.previewed': 'operator.plan.previewed',
  'plan.confirmed': 'operator.plan.confirmed',
  'plan.confirmation_rejected': 'operator.plan.confirmation_rejected',
  'plan.cancelled': 'operator.plan.cancelled',
  'plan.expired': 'operator.plan.expired',
  'plan.execution_started': 'operator.execution.started',
  'plan.execution_completed': 'operator.execution.completed',
  'plan.execution_blocked': 'operator.execution.blocked',
  'plan.execution_failed': 'operator.execution.failed',
  'plan.reconciliation_required': 'operator.execution.reconciliation_required',
  'workflow.approval_decided': 'workflow.approval.decided',
}) satisfies Readonly<Record<GovernedPlanAuditEvent['type'], CollaborationEventType>>;

interface DirectoryIdentity {
  readonly path: string;
  readonly real: string;
  readonly dev: bigint;
  readonly ino: bigint;
}

function directory(path: string, create: boolean): DirectoryIdentity {
  if (create && !existsSync(path)) mkdirSync(path, { recursive: false, mode: 0o700 });
  const symbolic = lstatSync(path, { bigint: true });
  const stat = statSync(path, { bigint: true });
  const real = realpathSync(path);
  if (
    symbolic.isSymbolicLink() ||
    !symbolic.isDirectory() ||
    !stat.isDirectory() ||
    real !== resolve(path)
  ) {
    throw new TypeError('Governed audit path must be a real canonical directory.');
  }
  return Object.freeze({ path, real, dev: stat.dev, ino: stat.ino });
}

function sameDirectory(identity: DirectoryIdentity): boolean {
  try {
    const symbolic = lstatSync(identity.path, { bigint: true });
    const stat = statSync(identity.path, { bigint: true });
    return (
      !symbolic.isSymbolicLink() &&
      symbolic.isDirectory() &&
      stat.isDirectory() &&
      realpathSync(identity.path) === identity.real &&
      stat.dev === identity.dev &&
      stat.ino === identity.ino
    );
  } catch {
    return false;
  }
}

interface SafeAuditRecord {
  readonly type: GovernedPlanAuditEvent['type'];
  readonly eventId: string;
  readonly occurredAt: string;
  readonly planId: string;
  readonly kind: string;
  readonly actorId: string;
  readonly workspaceId: string;
  readonly correlationId: string;
  readonly state: string;
  readonly revision: number;
  readonly evidenceRefs: readonly GovernedJsonValue[];
}

function auditRecord(value: GovernedPlanAuditEvent): SafeAuditRecord {
  const event = canonicalizeGovernedJson(value);
  if (!event || typeof event !== 'object' || Array.isArray(event)) {
    throw new TypeError('Governed audit event is invalid.');
  }
  const object = event as { readonly [key: string]: GovernedJsonValue };
  const allowed = new Set([
    'schema',
    'eventId',
    'type',
    'occurredAt',
    'planId',
    'kind',
    'actorId',
    'workspaceId',
    'correlationId',
    'state',
    'revision',
    'evidenceRefs',
    'details',
  ]);
  if (
    Object.keys(object).some((key) => !allowed.has(key)) ||
    object.schema !== 'openslack.governed_plan_audit.v1' ||
    typeof object.type !== 'string' ||
    !Object.hasOwn(EVENT_TYPES, object.type) ||
    typeof object.eventId !== 'string' ||
    typeof object.occurredAt !== 'string' ||
    !Number.isFinite(Date.parse(object.occurredAt)) ||
    typeof object.planId !== 'string' ||
    typeof object.kind !== 'string' ||
    typeof object.actorId !== 'string' ||
    typeof object.workspaceId !== 'string' ||
    typeof object.correlationId !== 'string' ||
    typeof object.state !== 'string' ||
    typeof object.revision !== 'number' ||
    !Number.isSafeInteger(object.revision) ||
    !Array.isArray(object.evidenceRefs)
  ) {
    throw new TypeError('Governed audit event failed its closed sink contract.');
  }
  return object as unknown as SafeAuditRecord;
}

export function createGovernedPlanCollaborationAuditSink(
  workspaceRootValue: string,
): GovernedPlanAuditSink {
  if (
    typeof workspaceRootValue !== 'string' ||
    resolve(workspaceRootValue) !== workspaceRootValue
  ) {
    throw new TypeError('Governed audit workspace root must be absolute and normalized.');
  }
  const root = directory(workspaceRootValue, false);
  const localPath = join(root.real, '.openslack.local');
  const local = directory(localPath, true);
  const collaboration = directory(join(local.real, 'collaboration'), true);
  const eventAppender = createBoundEventAppender(root.real);

  return async (value): Promise<void> => {
    if (!sameDirectory(root) || !sameDirectory(local) || !sameDirectory(collaboration)) {
      throw new TypeError('Governed audit directory identity changed.');
    }
    const event = auditRecord(value);
    const type = EVENT_TYPES[event.type];
    const objectKind: ObjectKind = event.type === 'workflow.approval_decided' ? 'workflow' : 'plan';
    const collaborationEvent: CollaborationEvent = {
      schema: 'openslack.collaboration_event.v1',
      id: event.eventId,
      timestamp: event.occurredAt,
      type,
      actor: {
        id: event.actorId,
        kind: event.type === 'workflow.approval_decided' ? 'human' : 'system',
      },
      object: { kind: objectKind, id: event.planId },
      source: { kind: 'operator', ref: 'qoder-governed-plan' },
      summary: `Recorded governed ${event.kind} transition ${event.type}.`,
      owner: {
        id: event.actorId,
        kind: event.type === 'workflow.approval_decided' ? 'human' : 'system',
      },
      severity:
        event.type === 'plan.execution_failed' || event.type === 'plan.reconciliation_required'
          ? 'warning'
          : 'info',
      visibility: 'workspace',
      correlationId: event.correlationId,
      redacted: true,
      containsSensitiveData: false,
      metadata: {
        auditEventId: event.eventId,
        workspaceId: event.workspaceId,
        state: event.state,
        revision: event.revision,
        evidenceRefs: event.evidenceRefs,
      },
    };
    const validation = validateEvent(collaborationEvent);
    const redaction = sanitizeEvent(collaborationEvent);
    if (!validation.valid || !redaction.safe) {
      throw new TypeError('Governed audit event failed Collaboration validation.');
    }
    eventAppender.append(collaborationEvent);
  };
}

File: services/governance-control/audit.go
package governancecontrol

import "github.com/Negentropy-Laby/OpenSlack/services/governance-control/internal/canonicaljson"

type AuditEvent struct {
	EventID       string
	Type          string
	OccurredAt    string
	PlanID        string
	Kind          string
	ActorID       string
	WorkspaceID   string
	CorrelationID string
	State         State
	Revision      int
	EvidenceRefs  []string
}

var auditEventTypes = map[string]struct{}{
	"plan.previewed": {}, "plan.confirmed": {}, "plan.confirmation_rejected": {},
	"plan.cancelled": {}, "plan.expired": {}, "plan.execution_started": {},
	"plan.execution_completed": {}, "plan.execution_blocked": {},
	"plan.execution_failed": {}, "plan.reconciliation_required": {},
	"workflow.approval_decided": {},
}

func ValidateAuditJSON(input []byte) (AuditEvent, error) {
	value, err := parseValue(input)
	if err != nil {
		return AuditEvent{}, err
	}
	root, ok := value.(canonicaljson.Object)
	if !ok {
		return AuditEvent{}, fail(ErrorInvalid, "$", "audit event must be an object.")
	}
	if err := exactKeys(root,
		[]string{"schema", "eventId", "type", "occurredAt", "planId", "kind", "actorId", "workspaceId", "correlationId", "state", "revision", "evidenceRefs"},
		[]string{"details"}, "$", "audit"); err != nil {
		return AuditEvent{}, err
	}
	schema, err := requiredString(root, "schema", nil, 64, "$", "schema")
	if err != nil || schema != "openslack.governed_plan_audit.v1" {
		if err != nil {
			return AuditEvent{}, err
		}
		return AuditEvent{}, fail(ErrorInvalid, "$/schema", "Governed plan audit schema is invalid.")
	}
	eventID, err := requiredString(root, "eventId", identifierPattern, 512, "$", "eventId")
	if err != nil {
		return AuditEvent{}, err
	}
	eventType, err := requiredString(root, "type", nil, 128, "$", "type")
	if err != nil {
		return AuditEvent{}, err
	}
	if _, exists := auditEventTypes[eventType]; !exists {
		return AuditEvent{}, fail(ErrorInvalid, "$/type", "audit event type is invalid.")
	}
	occurredAt, err := requiredTimestamp(root, "occurredAt", "$")
	if err != nil {
		return AuditEvent{}, err
	}
	planID, err := requiredString(root, "planId", planIDPattern, 512, "$", "planId")
	if err != nil {
		return AuditEvent{}, err
	}
	kind, err := requiredString(root, "kind", kindPattern, 512, "$", "kind")
	if err != nil {
		return AuditEvent{}, err
	}
	actorID, err := requiredString(root, "actorId", identifierPattern, 512, "$", "actorId")
	if err != nil {
		return AuditEvent{}, err
	}
	workspaceID, err := requiredString(root, "workspaceId", identifierPattern, 512, "$", "workspaceId")
	if err != nil {
		return AuditEvent{}, err
	}
	correlationID, err := requiredString(root, "correlationId", identifierPattern, 512, "$", "correlationId")
	if err != nil {
		return AuditEvent{}, err
	}
	stateText, err := requiredString(root, "state", nil, 64, "$", "state")
	if err != nil {
		return AuditEvent{}, err
	}
	state := State(stateText)
	if _, exists := transitions[state]; !exists {
		return AuditEvent{}, fail(ErrorInvalid, "$/state", "state is invalid.")
	}
	revision, err := requiredInteger(root, "revision", 1, 9_007_199_254_740_991, "$")
	if err != nil {
		return AuditEvent{}, err
	}
	rawRefs, err := requiredArray(root, "evidenceRefs", "$")
	if err != nil {
		return AuditEvent{}, err
	}
	refs := make([]string, len(rawRefs))
	for index, raw := range rawRefs {
		text, ok := raw.(string)
		if !ok || text == "" || len([]byte(text)) > MaxEvidenceRefBytes {
			return AuditEvent{}, fail(ErrorInvalid, "$/evidenceRefs", "evidenceRef is invalid.")
		}
		refs[index] = text
	}
	return AuditEvent{
		EventID: eventID, Type: eventType, OccurredAt: occurredAt, PlanID: planID,
		Kind: kind, ActorID: actorID, WorkspaceID: workspaceID,
		CorrelationID: correlationID, State: state, Revision: revision, EvidenceRefs: refs,
	}, nil
}
